### ISSUE TITLE
Mixed precision infrastructure lib

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#1979]
While #1978 updates the infrastructure to support 32 and 64 bit fields, the default definition of `r_solver` is still 64 bit:
```
  ! Default real kind for r_solver.
#if (R_SOLVER_PRECISION == 32)
  integer,      parameter :: r_solver = real32
  character(3), parameter :: PRECISION_R_SOLVER = '32'
#elif (R_SOLVER_PRECISION == 128)
  integer,      parameter :: r_solver = real128
  character(3), parameter :: PRECISION_R_SOLVER = '128'
#else
  integer,      parameter :: r_solver = real64
  character(3), parameter :: PRECISION_R_SOLVER = '64'
#endif
```
since no R_SOLVER_PRECISION is provided in the Makefile.

We need to investigate if we can change this to 32 bit, so we really test mixed precision.

**Comment by sergisiso** [comment]
Yes, I changed the value in #1968 and been using it for a while. I works well. The PR is ready (when the CI python 3.6 passes)

**Comment by hiker** [comment]
Hmm - I think then we can just close this ticket. Your PR #1968 will enable mixed precision, then #1978 means the compilation test of lfric/eg19 will then indeed be done with 32- and 64-bit.

**Comment by hiker** [comment]
Sorry, I have to reopen this. Merging in #1978 will allow 32-bit r_solver, 64-bit r_def, but the default values defined in `constants_mod.F90` are still set to be 64/64 (as pointed out in the code above).

I just tried:
```
--- a/src/psyclone/tests/test_files/dynamo0p3/infrastructure/utilities/constants_mod.F90
+++ b/src/psyclone/tests/test_files/dynamo0p3/infrastructure/utilities/constants_mod.F90
@@ -96,8 +96,8 @@ module constants_mod
   integer,      parameter :: r_solver = real128
   character(3), parameter :: PRECISION_R_SOLVER = '128'
 #else
-  integer,      parameter :: r_solver = real64
-  character(3), parameter :: PRECISION_R_SOLVER = '64'
+  integer,      parameter :: r_solver = real32
+  character(3), parameter :: PRECISION_R_SOLVER = '32'
 #endif
 
```
This should mean that even at compile time `r_solver` is 32-bit, so we test the real thing. I just tried on #1978, and all tests, examples and tutorials compiled as expected, but we need to make the above change (which is why I reopen this ticket).


The change @sergisiso mentioned (32 bit for r_solver), was I guess the following setting in `LFRicConstants`:
```
        LFRicConstants.PRECISION_MAP = {
...
                                           "r_solver": 4,

```

Do we actually need this information in PSyclone? As far as I could see the keys (names of precisions) are used frequently, and only one case where the actual size is used (in dynamo0p3.py):
```
actual_precision = LFRicConstants.PRECISION_MAP[actual_precision.name]
...
alg_precision = LFRicConstants.PRECISION_MAP[alg_arg.precision]
if alg_precision != actual_precision:
...
```
Do we need the byte values here, or can we just compare the names?

**Comment by hiker** [comment]
I also just realised that lfric example 19 is missing in the User Guide. Since eg19 is mixed precision, this user guide should als be quickly updated as part of this ticket.
